pkg/controller/volume/attachdetach/reconciler/reconciler.go

```diff
@@ -279,11 +291,21 @@ func (rc *reconciler) attachDesiredVolumes() {
 			}
 		}
 
+		// insert a attach operation start time into cache if needed for the volume for metrics reporting
+		// note that the plugin name has been set to "" as there is no visibility at this moment to which
+		// plugin/csi driver would be conducting the operation. It will be populated later when detach
+		// operation function is created and plugin name then could be decided
+		// NOTE: the metric does NOT include the wait time of potential "detach" operation might be introduced in "IsOperationPending" call
+		//       Including the "detach" time might introduce challenges to define SLO or to set alerts based on it.
+		//       Whether the "detach" operation will happen depends on the current status of the volume (currently attached to any none-desired node or not)
+		//       and "detach" can be a long operation.
+		rc.operationStartTimeCache.AddIfNotExist(operationexecutor.CreateAttachDetachOperationUniqueKey(volumeToAttach.VolumeName, volumeToAttach.NodeName), "", "volume_attach")
```

Comment:
same thing here, can we start the timer once we know the plugin? The logic will be simpler to follow if everything is initialized at the same time.

Reply:
we could certainly do that as @davidz627 mentioned previously [here](https://github.com/kubernetes/kubernetes/pull/80832#discussion_r314907693)
I prefer to put in to reconciler instead just for the sake of covering "waiting_time" in the future, such that the cache management happens within the same unit.
The other reason, which does not hold for now but might hold later, is for error metric reporting. Operation generator returns error in many cases (missing plugin could be one of it), when should it add the start timestamp into cache? If put after a valid plugin is found, the cache will not have the entry and makes error metric reporting tricky.
thoughts?

Reply:
Right now cache management is not in the same component. We're starting it in reconciler, and finishing it in operation generator.

I think we'll need to revisit error reporting/categorization.  Plugin-level errors can be tracked per operation, and we can get error percentage out of that.  I don't think higher level errors at the controller level should be tracked with the same metric, as there's not a great notion of error percentage.

Reply:
when I say management, I actually meant adding into cache for "waiting_time". the deletion along with metric reporting needs to happen in operation generator because of various of reasons.
yeah, error reporting/categorization needs to be planned out before it is added for sure.

Reply:
Hm that means for the metric that measures time waiting for detach before attaching, the plugin name will be empty? Is that odd? @gnufied 

I would rather keep this logic as simple as possible and not have a possibility of plugin name to be empty, and have to make 2 calls in 2 different places to actually initialize the metric.

Reply:
It will be odd indeed to emit the metric without the plugin name for "waiting_for_detach" case. The only reason we do not have plugin name here is because we have not actually inspected the volume right? This can be fixed later I think.

Reply:
@gnufied I have updated the PR according to the feedback, anyways I am not including "waiting_for_detach" within this PR.